### PR TITLE
Shoutcast autoreconnect feature

### PR DESCRIPTION
**Pull request by illuusio**
This implements Shoutcast autoreconnect feature. It's BETA code and why I make pull request is I can get more testers for this. It survives server not connectible on start and one time disconnection from server. More of these should be ironed out and see if this implementation fits or should it be changed somehow.

### CODE REVIEW

File: src/engine/sidechain/engineshoutcast.cpp [hunk 1]
```diff
@@ -198,6 +200,10 @@ void EngineShoutcast::updateFromPreferences() {
     m_ogg_dynamic_update = (bool)m_pConfig->getValueString(
             ConfigKey(SHOUTCAST_PREF_KEY,"ogg_dynamicupdate")).toInt();
 
+    // Dynamic Ogg metadata update
```
Comment (ywwg) [inline src/engine/sidechain/engineshoutcast.cpp]: what does this have to do with ogg?

File: src/engine/sidechain/engineshoutcast.cpp [hunk 2]
```diff
@@ -533,13 +558,39 @@ bool EngineShoutcast::writeSingle(const unsigned char* data, size_t len) {
...
     // If we are here then the user wants to be connected (shoutcast is enabled
     // in the preferences).
 
-    // If we aren't connected, bail.
-    if (m_iShoutStatus != SHOUTERR_CONNECTED)
+    // If we aren't connected, bail and autoreconnect is off.
+    if (m_iShoutStatus != SHOUTERR_CONNECTED
+        && !m_stream_autoreconnect) {
         return;
+    } else if (m_iShoutStatus != SHOUTERR_CONNECTED
```
Comment (ywwg) [inline src/engine/sidechain/engineshoutcast.cpp]: no need for "else" since you have a return statement.  Just start a new if()

File: src/engine/sidechain/engineshoutcast.cpp [hunk 3]
```diff
@@ -533,13 +558,39 @@ bool EngineShoutcast::writeSingle(const unsigned char* data, size_t len) {
...
+               && m_stream_autoreconnect
+               && getState() != NETWORKSTREAMWORKER_STATE_CONNECTING) {
+        bool l_bConnectionSuccess = false;
+        // Trying to connect to server
+        // If autoreconnect is on we just keep trying until
+        // the end of universe or Shoutcast is not enabled anymore
+        while (true) {
```
Comment (ywwg) [inline src/engine/sidechain/engineshoutcast.cpp]: while (!l_bConnectionSuccess)?

File: src/engine/sidechain/engineshoutcast.cpp [hunk 4]
```diff
@@ -732,12 +784,24 @@ void EngineShoutcast::run() {
...
-    if (!processConnect()) {
-        errorDialog(tr("Can't connect to streaming server"),
-                    tr("Please check your connection to the Internet and verify that your username and password are correct."));
-        return;
-    }
 
+    // Trying to connect to server
+    // If autoreconnect is on we just keep trying until
+    // the end of universe or Shoutcast is not enabled anymore
+    while (true) {
```
Comment (ywwg) [inline src/engine/sidechain/engineshoutcast.cpp]: this is duplicated code.  Refactor into a function

File: src/engine/sidechain/engineshoutcast.cpp (hunk 3 above)
Comment (daschuer) [inline src/engine/sidechain/engineshoutcast.cpp]: We may also check for the "enabled" CO in the loop. This way we can continue very long, without delay the exit. 
The same for the other loop.

By the way, cam one confirm the deadlock or do we hunt a phantom?

File: src/engine/sidechain/engineshoutcast.cpp (hunk 1 above)
Comment (illuusio) [inline src/engine/sidechain/engineshoutcast.cpp]: That I would also know.. Even code is very trivial and if I understood correct changing Ogg stream metadata was problem in some libshout and if you don't tap that checkbox metadata won't be updated.

**Comment by daschuer** [comment]
Plain 1.12 branch: 

Yesterday, I have experienced an issue, that the stream has broken. I got the error meassage box and the  disconnected message box.  I was not able to reconnect manually. The shoutcast status sticks on error stat and it doesn't matter if I change the enable state. I was immediately able to reconnect after restart of Mixxx. 

I did expect to go at least to the connecting state after toggling "enabled", but I did not see that. 

After that I was trying to debug the issue, but I was not able to reproduce.

**Comment by illuusio** [comment]
Sorry to say but there probably lot's of these. Connection code is little bit messy because of nature of trying to solve every problem. Did it print anything after it didn't connect?

**Comment by daschuer** [comment]
I will have a look if I can find the log file tonight.

**Comment by rryan** [comment]
Hello! This branch is against the 1.12 branch, which is now closed to any changes except bugfixes. Please re-open your PR against master if this PR is still relevant.
